### PR TITLE
[bitnami/etcd] feat: allow reuse startFromSnapshot volume in the disasterRecovery

### PR DESCRIPTION
**Pull request by relzhong**
add support to allow reuse startFromSnapshot volume in the disasterRecovery. fix breaking change to relate [pr](https://github.com/bitnami/charts/pull/4940)
relate pr in bitnami-docker-etcd [pr](https://github.com/bitnami/bitnami-docker-etcd/pull/22)

**Benefits**

**Possible drawbacks**

**Applicable issues**

  - fixes #

**Additional information**

**Checklist** 

- [X] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [X] Variables are documented in the README.md
- [X] Title of the PR starts with chart name (e.g. [bitnami/<name_of_the_chart>])

### CODE REVIEW

**Comment by yilmi** [comment]
Let's wait for the container changes from https://github.com/bitnami/bitnami-docker-etcd/pull/22 are reviewed an merge before working on this one 👍

**Comment by yilmi** [comment]
@relzhong reviewing your change and the description from https://github.com/bitnami/charts/pull/4940, I don't see what problem you're trying to solve.

Could you please provide more details in your issue, such as Benefits, Possible Drawbacks and more details to explain why the current logic doesn't work for you?

Thanks,
Yass

**Comment by relzhong** [comment]
Hi @yilmi ,
In current disasterRecovery phase, we need to starting a new cluster using an existing snapshot, so the precedure will like this:
1. create cluster and add a disasterrecovery pvc.
```
  --disasterRecovery.pvc.existingClaim=pvc1
```
2. after first snapshot, we get snapshot1, then we uninstall the chart. and recreate using start from snapshot
```
  --startFromSnapshot.existingClaim=pvc1 --startFromSnapshot.snapshotFilename=snapshot1
```
if we also need disasterRecovery, we should create pvc2, and set params to 
```
--disasterRecovery.pvc.existingClaim=pvc2 --startFromSnapshot.existingClaim=pvc1 --startFromSnapshot.snapshotFilename=snapshot1 
```
the reason why create pvc2 is beacuse setting ETCD_INIT_SNAPSHOT_DIR to a fix /init-snapshot.

with this pr, we can just modify to ETCD_INIT_SNAPSHOT_DIR to /snapshots. so we just need one pvc pvc1. and bring back cluster will be this 
```
--disasterRecovery.pvc.enabled=true --startFromSnapshot.existingClaim=pvc1 --startFromSnapshot.snapshotFilename=snapshot1 
```
The advantage is we need just one pvc and no need to switch pvc between pvc1 and pv2 on every disaster recovery.

Thanks.

**Comment by yilmi** [comment]
Hi @relzhong, thanks for providing more details!

I read through the PR you mention in your description, and I believe that your desired state would be achieved by the case 4.2 described from https://github.com/bitnami/charts/pull/4940

> So only in the case of
>
> `disasterRecovery.enabled = true` && `startFromSnapshot.enabled = true` && ! `disasterRecovery.pvc.existingClaim`
>
> it is when the volume passed as `startFromSnapshot.existingClaim` would be used to make the function of the one that we would create for the disasterRecovery

The way I read it, if you set:
- `disasterRecovery.enabled=true`
- `startFromSnapshot.enabled=true`
- `startFromSnapshot.existingClaim="xxxx"`

Without setting `disasterRecovery.enabled`, you should get the behaviour you want.

**Comment by relzhong** [comment]
In #4940, the logic had been merge to master. However, after the [refactor](https://github.com/bitnami/charts/pull/5682) of both etcd chart and  bitnami-docker-etcd, the logic had been move from chart to docker-etcd. The [script](https://github.com/bitnami/charts/blob/1fb5d2c9a0473f74f11233b59b784982df34a9e3/bitnami/etcd/templates/scripts-configmap.yaml) changed in #4940 cannot be found in current master branch. But the script in bitnami-docker-etcd has a fixed directory /init-snapshot. So i modify both project to achieve this.

**Comment by yilmi** [comment]
Hi @relzhong, thanks for the pointers. Let me take some time to test and look into this before getting back to you. I should be able to do it by the end of the week or early next week. 👍

**Review by juan131 — CHANGES_REQUESTED** [review]
I agree with @yilmi , the idea off [this PR ](https://github.com/bitnami/charts/pull/4940) was reusing the same PV both for "start from snapshot" and "disaster recovery features"... But it's true that we messed up this when we moved the logic to the container as you pointed...

**Comment by juan131** [comment]
Potential fix for https://github.com/bitnami/charts/issues/6251

**Comment by relzhong** [comment]
fixed @juan131

File: bitnami/etcd/templates/statefulset.yaml [hunk 1]
```diff
@@ -217,6 +217,10 @@ spec:
             - name: ETCD_INIT_SNAPSHOT_FILENAME
               value: {{ .Values.startFromSnapshot.snapshotFilename | quote }}
             {{- end }}
+            {{- if and .Values.disasterRecovery.enabled .Values.startFromSnapshot.enabled (not .Values.disasterRecovery.pvc.existingClaim) }}
+            - name: ETCD_INIT_SNAPSHOTS_DIR
+              value: "/snapshots"
```
Comment (juan131) [inline bitnami/etcd/templates/statefulset.yaml:222]:
```suggestion
            - name: ETCD_INIT_SNAPSHOTS_DIR
              value: {{ ternary "/snapshots" "/init-snapshot" (and .Values.disasterRecovery.enabled (not .Values.disasterRecovery.pvc.existingClaim)) | quote }}
```

Comment (juan131) [reply bitnami/etcd/templates/statefulset.yaml:222]: What do you think about this proposal? This way we always align `ETCD_INIT_SNAPSHOTS_DIR` with the "volumeMounts" used in the chart (regardless the default value used in the image).

Comment (relzhong) [reply bitnami/etcd/templates/statefulset.yaml]: Sure, it's better solution

**Review by juan131 — APPROVED** [review]
LGTM

**Comment by juan131** [comment]
You can always build your custom image. To do so, clone the https://github.com/bitnami/bitnami-docker-etcd repo, browse to the directory `3/debian-10`, and apply the changes below in the Dockerfile:
 
```diff
 FROM docker.io/bitnami/minideb:buster
 (...)
 RUN . /opt/bitnami/scripts/libcomponent.sh && component_unpack "gosu" "1.14.0-0" --checksum 3e6fc37ca073b10a73a804d39c2f0c028947a1a596382a4f8ebe43dfbaa3a25e
-RUN . /opt/bitnami/scripts/libcomponent.sh && component_unpack "etcd" "3.5.1-0" --checksum 912653e5c0879fd27c8bb0327bea171d7ae504611ec0a9c8fcc2ec6184766a63
+RUN . /opt/bitnami/scripts/libcomponent.sh && component_unpack "etcd" "3.4.16-0" --checksum 87ffa91cbaf5ad953cae045b8dfb4cb4f12c66da499bd332d2b4acf2f01941f0
 RUN chmod g+rwX /opt/bitnami
(...)
 ENV BITNAMI_APP_NAME="etcd" \
-    BITNAMI_IMAGE_VERSION="3.5.1-debian-10-r4" \
+    BITNAMI_IMAGE_VERSION="3.4.16-debian-10-r0" \
     ETCDCTL_API="3"
(...)
 CMD [ "/opt/bitnami/scripts/etcd/run.sh" ]
```

Then, use `docker build ...` to build your custom etcd image that uses `3.4.16` but including the latest scripts.

**Pull request merged**